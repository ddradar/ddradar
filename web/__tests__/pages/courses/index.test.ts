import { testCourseData as course } from '@ddradar/core/test/data'
import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import { ref } from 'vue'

import { mountAsync } from '~/__tests__/test-utils'
import Page from '~/pages/courses/index.vue'

describe('Page /courses', () => {
  const query = { type: '1', series: '16' }
  const courses = [
    {
      id: course.id,
      name: course.name,
      series: course.series,
      charts: course.charts.map(c => ({
        playStyle: c.playStyle,
        difficulty: c.difficulty,
        level: c.level,
      })),
    },
  ]

  describe('snapshot tests', () => {
    test('{ isLoading: true } renders loading state', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ query } as any)
      vi.mocked(useFetch).mockResolvedValue({
        pending: ref(true),
        data: ref([]),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig]],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
    test('{ isLoading: false, courses: <Data> } renders course list', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ query } as any)
      vi.mocked(useFetch).mockResolvedValue({
        pending: ref(false),
        data: ref(courses),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig]],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
    test('{ isLoading: false, courses: [] } renders no data', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ query } as any)
      vi.mocked(useFetch).mockResolvedValue({
        pending: ref(false),
        data: ref([]),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig]],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  test.each([
    [undefined, undefined, 'COURSES'],
    [undefined, '16', 'COURSES (A20)'],
    ['1', '17', 'NONSTOP (A20 PLUS)'],
    ['2', '18', '段位認定 (A3)'],
  ])('?type=%s&series=%s renders "%s" title', async (type, series, title) => {
    // Arrange
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useRoute).mockReturnValue({ query: { type, series } } as any)
    vi.mocked(useFetch).mockResolvedValue({
      pending: ref(true),
      data: ref([]),
    } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */

    // Act
    const wrapper = await mountAsync(Page, {
      global: {
        plugins: [[Oruga, bulmaConfig]],
        stubs: { NuxtLink: RouterLinkStub },
      },
    })

    // Assert
    expect(wrapper.find('h1').text()).toBe(title)
  })
})
