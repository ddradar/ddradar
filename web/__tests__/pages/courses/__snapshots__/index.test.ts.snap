// Vitest Snapshot v1

exports[`Page /courses > snapshot tests > { isLoading: false, courses: <Data> } renders course list 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    NONSTOP (A20)
  </h1>
  <div
    class="buttons"
  >
    
    <a
      class="button is-info is-outlined"
      disabled="true"
    >
      NONSTOP (A20)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A20)
    </a>
    <a
      class="button is-info"
    >
      NONSTOP (A20 PLUS)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A20 PLUS)
    </a>
    <a
      class="button is-info"
    >
      NONSTOP (A3)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A3)
    </a>
    
  </div>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Series
      </span>
      <span
        data-id="2"
      >
        Name
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr
            class=""
            draggable="false"
          >
            <!--v-if-->
            <!--v-if-->
            
            <td
              class=""
              data-label="Series"
            >
              A20
            </td>
            <td
              class=""
              data-label="Name"
            >
              <a>
                TWENTY
              </a>
            </td>
            
            <!--v-if-->
          </tr>
          <transition-stub
            appear="false"
            css="true"
            name=""
            persisted="false"
          >
            <!--v-if-->
          </transition-stub>
          <!--v-if-->
          
          
          <!--v-if-->
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub
        appear="false"
        css="true"
        name="fade"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination is-right"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              <li>
                <a
                  aria-current="true"
                  class="pagination-link is-current"
                  href="#"
                  role="button"
                >
                  
                  1
                  
                </a>
              </li>
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;

exports[`Page /courses > snapshot tests > { isLoading: false, courses: [] } renders no data 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    NONSTOP (A20)
  </h1>
  <div
    class="buttons"
  >
    
    <a
      class="button is-info is-outlined"
      disabled="true"
    >
      NONSTOP (A20)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A20)
    </a>
    <a
      class="button is-info"
    >
      NONSTOP (A20 PLUS)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A20 PLUS)
    </a>
    <a
      class="button is-info"
    >
      NONSTOP (A3)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A3)
    </a>
    
  </div>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Series
      </span>
      <span
        data-id="2"
      >
        Name
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped is-empty"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr>
            <td
              colspan="2"
            >
              
              <section
                class="section"
              >
                <div
                  class="content has-text-grey has-text-centered"
                >
                  <p>
                    No Data
                  </p>
                </div>
              </section>
              
            </td>
          </tr>
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub
        appear="false"
        css="true"
        name="fade"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination is-right"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;

exports[`Page /courses > snapshot tests > { isLoading: true } renders loading state 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    NONSTOP (A20)
  </h1>
  <div
    class="buttons"
  >
    
    <a
      class="button is-info is-outlined"
      disabled="true"
    >
      NONSTOP (A20)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A20)
    </a>
    <a
      class="button is-info"
    >
      NONSTOP (A20 PLUS)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A20 PLUS)
    </a>
    <a
      class="button is-info"
    >
      NONSTOP (A3)
    </a>
    <a
      class="button is-info"
    >
      段位認定 (A3)
    </a>
    
  </div>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Series
      </span>
      <span
        data-id="2"
      >
        Name
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped is-empty"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr>
            <td
              colspan="2"
            >
              
              <section
                class="section"
              >
                <div
                  class="b-skeleton is-animated"
                >
                  <div
                    class="b-skeleton-item is-rounded"
                  />
                  <div
                    class="b-skeleton-item is-rounded"
                  />
                  <div
                    class="b-skeleton-item is-rounded"
                  />
                </div>
              </section>
              
            </td>
          </tr>
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub
        appear="false"
        css="true"
        name="fade"
        persisted="false"
      >
        <div
          class="o-load loading"
        >
          <div
            class="o-load__overlay loading-overlay"
          />
          
          <span
            class="icon is-spin is-medium o-load__icon icon"
          >
            <i
              class="mdi mdi-loading mdi-36px"
            />
          </span>
          
        </div>
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination is-right"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;
